oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticBulkProcessorHandler.java

```diff
@@ -75,6 +76,11 @@
      */
     private final Phaser phaser = new Phaser(1); // register main controller
 
+    /**
+     * Throwable object containing error/exception which occurred while trying to update index in elasticsearch.
+     */
+    private Throwable throwable;
```

Comment:
I would change this from `Throwable` to `ConnectionException` since we are checking the type (and casting it) in the `afterBulk`. This would simplify things also in `ElasticIndexWriter#close` (no need to try-catch since a `ConnectionException` extends from `IOException`).

Moreover, since this variable gets accessed by different threads you should make it `volatile` or use an `AtomicReference`.

Reply:
Oh, actually i forgot to remove the type check and casting in the `afterBulk` method. I am thinking that we should throw every `Throwable` we receive in `afterBulk` and let indexing happen again. WDYT?

Reply:
I would use an IOException (or some other checked or runtime exception that makes sense in this context), and wrap other exceptions / errors: new IOException(throwable)

Reply:
In general, I would prefer to not have Throwables in method's signatures since it would include also `Error`s. I understand that you used it because `BulkProcessor.Listener#afterBulk` has Throwable in its signature. I don't really understand why they took this decision: `afterBulk` is always invoked with Exceptions and never with Throwables (see [here](https://github.com/elastic/elasticsearch/blob/f72b893fd313810c4d6bf98585d64bc41958b976/server/src/main/java/org/elasticsearch/action/bulk/BulkRequestHandler.java#L54-L92) for reference).

What about specifying in the signature our own Exception (something like `ElasticBulkException`) with the Throwable as the cause? You should then try-catch it in the `IndexWriter`: if the cause is an IOException you throw the cause, otherwise you throw it wrapped as a `RuntimeException`.

Reply:
Yes I understand. If `close` method throws `Thwowable` and its user catches `Throwable`, it would be catching everything even if it doesn't want to. Ideally it should only catch Throwable if it has originated from elastic api.

I am working on removing `Throwable` from the method signature. I think even wrapping the `Throwable` in `IOException` would make sense.